reframe/core/pipeline.py

```diff
@@ -762,55 +766,70 @@ def param_space_len(cls):
 
     @classmethod
     def prepare_param_space(cls):
```

Comment:
Actually you don't need this. Here is the trick to create a generator from another one:

```python
    @classmethod
    def walk_param_space(cls):
        yield from itertools.product(*(p for p in cls._rfm_params.values()))
```

When you want to chain generators, you need the `yield from`, not just `yield`. So I think this will simplify the design in this part. You will only need to call `walk_param_space()` whenever you need to iterate over the parameter space.

Reply:
As discussed before, this is still needed by the `simple_test` decorator. A call to `prepare_param_space` will create the iterator that feeds the values to the parameters as the instances of the class get created. If this `prepare_param_space` method is not called before class instantiation, or the class is instantiated after the iterator has been exhausted, the test parameters will be set to `None`.